crate/operator/restore_backup.py

```diff
@@ -0,0 +1,925 @@
+            core = CoreV1Api(api_client)
+            data = await get_source_backup_repository_data(
+                core,
+                namespace,
+                name,
+                logger,
+            )
+            password, host = await asyncio.gather(
+                get_system_user_password(core, namespace, name),
+                get_host(core, namespace, name),
+            )
+            conn_factory = connection_factory(host, password)
+
+            await self._create_backup_repository(conn_factory, repository, data, logger)
+
+            await self._ensure_snapshot_exists(
+                conn_factory, repository, snapshot, logger
+            )
+
+            await self._start_restore_snapshot(
+                conn_factory, repository, snapshot, tables, logger
+            )
+
+    @staticmethod
+    async def _create_backup_repository(
+        conn_factory,
+        repository: str,
+        data: dict,
+        logger: logging.Logger,
+    ):
+        """
+        Create a backup repository with the given credentials if it does not exist yet.
+
+        :param conn_factory: A function that establishes a database connection to
+            the CrateDB cluster used for SQL queries.
+        :param repository: The name of the repository to be created.
+        :param data: a dict containing the bucket name, base path and secrets to access
+            the source backup repository.
+        :param logger: the logger on which we're logging
+        """
+        try:
+            async with conn_factory() as conn:
+                async with conn.cursor() as cursor:
+                    await cursor.execute(
+                        "SELECT * FROM sys.repositories WHERE name=%s", (repository,)
+                    )
+                    row = await cursor.fetchone()
+                    if not row:
+                        repository_ident = quote_ident(repository, cursor._impl)
+                        try:
+                            await cursor.execute(
+                                f"CREATE REPOSITORY {repository_ident} type s3 with "
+                                "(access_key = %s, secret_key = %s, bucket = %s, "
+                                "base_path = %s, readonly=true)",
+                                (
+                                    data["accessKeyId"],
+                                    data["secretAccessKey"],
+                                    data["bucket"],
+                                    data["basePath"],
+                                ),
+                            )
+                        except ProgrammingError as e:
+                            logger.warning(
+                                "CREATE REPOSITORY query failed.", exc_info=e
+                            )
+        except (DatabaseError, OperationalError) as e:
+            logger.warning("DatabaseError in _create_backup_repository: %s", str(e))
+            if "Unable to verify the repository" in str(e):
```

Comment:
Unfortunately I didn't find another way.. `e.pgcode` is always `XX000` for `internal_error` and `e.pgerror` sth like
```
[restore_backup_12345] Unable to verify the repository, [restore_backup_12345] is not accessible on master node: AmazonS3Exception 'The request signature we calculated does not match the signature you provided. Check your key and signing method. (Service: Amazon S3; Status Code: 403; Error Code: SignatureDoesNotMatch; Request ID: XXX; S3 Extended Request ID: XXX; Proxy: null)
```
with lots of different `Error Code` possibilities from AWS, only `Unable to verify the repository` was always the same.
But maybe we should just raise a `PermanentError` without further checks, if any of these queries fails? :thinking:

Reply:
> But maybe we should just raise a PermanentError without further checks, if any of these queries fails? 🤔

I think so, because I can't imagine any cases where it could fail temporarily 🤔

Reply:
ok, I raise a `PermanentError` now if anything goes wrong in one of the repository/snapshot queries and removed the checks for certain strings in the error msg.